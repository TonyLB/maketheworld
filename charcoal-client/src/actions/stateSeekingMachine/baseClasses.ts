interface ISSMWrappedAction<D extends Record<string, any>> {
    (dispatch: any, getState: any): Promise<Partial<D>>;
}

export interface ISSMAction<D extends Record<string, any>> {
    (data: D): ISSMWrappedAction<D>
}

export interface ISSMChoiceState<K> {
    key: K;
    stateType: 'CHOICE';
    choices: Array<K>;
}

export interface ISSMAttemptState<K, D extends Record<string, any>> {
    key: K;
    stateType: 'ATTEMPT';
    action: ISSMAction<D>;
    resolve: K;
    reject: K;
}

export interface ISSMHoldCondition<D extends Record<string, any>> {
    (data: D, getState: any): Boolean;
}

export interface ISSMHoldState<K, D extends Record<string, any>> {
    key: K;
    stateType: 'HOLD';
    condition: ISSMHoldCondition<D>;
    next: K;
}

export type ISSMPotentialState<K, D extends Record<string, any>> = ISSMChoiceState<K> | ISSMAttemptState<K, D> | ISSMHoldState<K, D>

export interface ISSMTemplateAbstract<K extends string, D extends Record<string, any>> {
    initialState: K;
    initialData: D;
    states: Record<K, ISSMPotentialState<K, D>>;
}

export interface IStateSeekingMachineAbstract<K extends string, D extends Record<string, any>, T extends ISSMTemplateAbstract<K, D>> {
    key: string;
    template: T;
    currentState: K;
    desiredState: K;
    data: D;
}
