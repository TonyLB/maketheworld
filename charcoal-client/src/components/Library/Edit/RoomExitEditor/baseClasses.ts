export type RoomExit = {
    key: string;
    to: string;
    from: string;
    name: string;
}
