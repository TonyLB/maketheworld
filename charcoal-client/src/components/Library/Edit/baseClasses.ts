import { BaseEditor, Selection } from 'slate'
import { ReactEditor } from 'slate-react'

export type CustomText = {
    text: string;
}

export type CustomLineElement = {
    type: 'line';
    children: CustomText[]
}

export type CustomActionLinkElement = {
    type: 'actionLink';
    to: string;
    key: string;
    children: CustomText[]
}

export type CustomFeatureLinkElement = {
    type: 'featureLink';
    to: string;
    key: string;
    children: CustomText[]
}

export type CustomLinkElement = CustomActionLinkElement | CustomFeatureLinkElement

export type CustomLineBreak = {
    type: 'lineBreak';
}

export type CustomParagraphContents = CustomText | CustomActionLinkElement | CustomFeatureLinkElement | CustomLineBreak

export const isCustomLineBreak = (item: CustomParagraphContents): item is CustomLineBreak => ('type' in item && item.type === 'lineBreak')
export const isCustomActionLink = (item: CustomParagraphContents): item is CustomActionLinkElement => ('type' in item && item.type === 'actionLink')
export const isCustomFeatureLink = (item: CustomParagraphContents): item is CustomFeatureLinkElement => ('type' in item && item.type === 'featureLink')
export const isCustomLink = (item: CustomParagraphContents): item is CustomLinkElement => (isCustomActionLink(item) || isCustomFeatureLink(item))
export const isCustomText = (item: CustomParagraphContents): item is CustomText => ('text' in item)

export type CustomParagraphElement = {
    type: 'paragraph';
    children: CustomParagraphContents[]
}

export type CustomDescriptionElement = {
    type: 'description';
    children: CustomParagraphElement[]
}

type CustomElement = CustomLineElement |
    CustomActionLinkElement |
    CustomFeatureLinkElement |
    CustomParagraphElement |
    CustomDescriptionElement



declare module 'slate' {
    interface CustomTypes {
      Editor: BaseEditor & ReactEditor & { saveSelection?: Selection }
      Element: CustomElement
      Text: CustomText
    }
  }

export {}