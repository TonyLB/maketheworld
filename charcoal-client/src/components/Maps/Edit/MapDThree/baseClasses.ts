import {
    SimulationNodeDatum,
    SimulationLinkDatum
} from 'd3-force'

export type SimNode = SimulationNodeDatum & {
    id: string;
    zLevel?: number;
    cascadeNode: boolean;
    roomId: string;
    visible: boolean;
}
export type NodeRecord = Record<string, SimNode>
export type LinkRecord = {
    id: string,
    source: string,
    target: string,
    visible?: boolean
}[]
export type SimulationReturn = {
    key: string,
    nodes: SimNode[],
    links: SimulationLinkDatum<SimNode>[]
}

export type SimCallback = (nodes: SimNode[]) => void

export type MapNodes = SimNode[]
export type MapLinks = SimulationLinkDatum<SimulationNodeDatum>[]
