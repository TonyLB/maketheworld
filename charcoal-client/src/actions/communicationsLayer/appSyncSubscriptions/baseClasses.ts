import { ISSMTemplateAbstract, ISSMPotentialState, ISSMAction, ISSMHoldCondition } from '../../stateSeekingMachine/baseClasses'

export const SUBSCRIPTION_SUCCESS = 'SUBSCRIPTION_SUCCESS'

export type subscriptionSSMKeys = 'INITIAL' | 'AVAILABLE' | 'SUBSCRIBING' | 'SUBSCRIBED' | 'SYNCHING' | 'SYNCHRONIZED' | 'UNSUBSCRIBING'
export type cachedSubscriptionSSMKeys = 'INITIAL' | 'FETCHING' | 'FETCHED' | 'SUBSCRIBING' | 'SUBSCRIBED' | 'SYNCHING' | 'SYNCHRONIZED' | 'UNSUBSCRIBING'

//
// Implement a state-seeking machine to keep websockets connected where possible.
//
export const subscriptionSSMClassGenerator = <D extends Record<string, any>, T extends string>({
    ssmType, initialData, condition, subscribeAction, unsubscribeAction, syncAction
}: {
    ssmType: T;
    initialData: D;
    condition: ISSMHoldCondition<D>;
    subscribeAction: ISSMAction<D>;
    unsubscribeAction: ISSMAction<D>;
    syncAction: ISSMAction<D>
}) => class implements ISSMTemplateAbstract<subscriptionSSMKeys, D> {
    ssmType: T = ssmType
    initialState: subscriptionSSMKeys = 'INITIAL'
    initialData: D = initialData
    states: Record<subscriptionSSMKeys, ISSMPotentialState<subscriptionSSMKeys, D>> = {
        INITIAL: {
            stateType: 'CHOICE',
            key: 'INITIAL',
            choices: ['AVAILABLE']
        },
        AVAILABLE: {
            stateType: 'HOLD',
            key: 'AVAILABLE',
            condition,
            next: 'SUBSCRIBING'
        },
        SUBSCRIBING: {
            stateType: 'ATTEMPT',
            key: 'SUBSCRIBING',
            action: subscribeAction,
            resolve: 'SUBSCRIBED',
            reject: 'INITIAL'
        },
        SUBSCRIBED: {
            stateType: 'CHOICE',
            key: 'SUBSCRIBED',
            choices: ['UNSUBSCRIBING', 'SYNCHING']
        },
        UNSUBSCRIBING: {
            stateType: 'ATTEMPT',
            key: 'UNSUBSCRIBING',
            action: unsubscribeAction,
            resolve: 'INITIAL',
            reject: 'INITIAL'
        },
        SYNCHING: {
            stateType: 'ATTEMPT',
            key: 'SYNCHING',
            action: syncAction,
            resolve: 'SYNCHRONIZED',
            reject: 'SUBSCRIBED'
        },
        SYNCHRONIZED: {
            stateType: 'CHOICE',
            key: 'SYNCHRONIZED',
            choices: ['UNSUBSCRIBING']
        },
    }
}

//
// Implement a more elaborate state-seeking machine to keep websockets connected where possible,
// including fetching cached information.
//
export const cachedSubscriptionSSMClassGenerator = <D extends Record<string, any>, T extends string>({
    ssmType, initialData, fetchAction, subscribeAction, unsubscribeAction, syncAction
}: {
    ssmType: T;
    initialData: D;
    fetchAction: ISSMAction<D>;
    subscribeAction: ISSMAction<D>;
    unsubscribeAction: ISSMAction<D>;
    syncAction: ISSMAction<D>
}) => class implements ISSMTemplateAbstract<cachedSubscriptionSSMKeys, D> {
    ssmType: T = ssmType
    initialState: cachedSubscriptionSSMKeys = 'INITIAL'
    initialData: D = initialData
    states: Record<cachedSubscriptionSSMKeys, ISSMPotentialState<cachedSubscriptionSSMKeys, D>> = {
        INITIAL: {
            stateType: 'CHOICE',
            key: 'INITIAL',
            choices: ['FETCHING']
        },
        FETCHING: {
            stateType: 'ATTEMPT',
            key: 'FETCHING',
            action: fetchAction,
            resolve: 'FETCHED',
            reject: 'INITIAL'
        },
        FETCHED: {
            stateType: 'CHOICE',
            key: 'FETCHED',
            choices: ['SUBSCRIBING']
        },
        SUBSCRIBING: {
            stateType: 'ATTEMPT',
            key: 'SUBSCRIBING',
            action: subscribeAction,
            resolve: 'SUBSCRIBED',
            reject: 'INITIAL'
        },
        SUBSCRIBED: {
            stateType: 'CHOICE',
            key: 'SUBSCRIBED',
            choices: ['UNSUBSCRIBING', 'SYNCHING']
        },
        UNSUBSCRIBING: {
            stateType: 'ATTEMPT',
            key: 'UNSUBSCRIBING',
            action: unsubscribeAction,
            resolve: 'INITIAL',
            reject: 'INITIAL'
        },
        SYNCHING: {
            stateType: 'ATTEMPT',
            key: 'SYNCHING',
            action: syncAction,
            resolve: 'SYNCHRONIZED',
            reject: 'SUBSCRIBED'
        },
        SYNCHRONIZED: {
            stateType: 'CHOICE',
            key: 'SYNCHRONIZED',
            choices: ['UNSUBSCRIBING']
        }
    }
}
