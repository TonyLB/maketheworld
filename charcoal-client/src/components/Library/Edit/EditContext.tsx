import React, { FunctionComponent, useContext } from "react"
import { GenericTreeNode } from "@tonylb/mtw-wml/dist/tree/baseClasses";
import { SchemaTag } from "@tonylb/mtw-wml/dist/schema/baseClasses";
import { TreeId } from "@tonylb/mtw-wml/dist/tree/baseClasses";

type EditContextType = {
    field: GenericTreeNode<SchemaTag, TreeId>;
    parentId: string;
    tag: 'ShortName' | 'Name' | 'Description' | 'Statement' | 'Fallthrough' | 'If'
}

const EditContext = React.createContext<EditContextType>({
    field: { data: { tag: 'Name' }, id: '', children: [] },
    parentId: '',
    tag: 'Name'
})

export const EditSchema: FunctionComponent<EditContextType> = ({ field, parentId, tag, children }) => {
    return <EditContext.Provider value={{ field, parentId, tag }}>
        { children }
    </EditContext.Provider>
}

export const useEditContext = () => (useContext(EditContext))
