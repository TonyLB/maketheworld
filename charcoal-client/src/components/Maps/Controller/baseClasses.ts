import { GenericTree } from "@tonylb/mtw-sequence/dist/tree/baseClasses"
import { SchemaConditionTag, SchemaExitTag, SchemaRoomTag } from "@tonylb/mtw-wml/dist/simpleSchema/baseClasses"

export type ToolSelected = 'Select' | 'Move' | 'AddRoom' | 'OneWayExit' | 'TwoWayExit'

export type MapTreeExit = SchemaExitTag & { inherited?: boolean }
export type MapTreeRoom = SchemaRoomTag & { inherited?: boolean }
export type MapTreeCondition = SchemaConditionTag & { inherited?: boolean }

export type MapTreeItem = MapTreeExit | MapTreeRoom | MapTreeCondition

type MapContextExitDrag = {
    sourceRoomId: string;
    x: number;
    y: number;
}

type MapDispatchSetTool = {
    type: 'SetToolSelected';
    value: ToolSelected;
}

type MapDispatchSetExitDrag = {
    type: 'SetExitDrag';
} & Partial<MapContextExitDrag>

export type MapDispatchAction = MapDispatchSetTool | MapDispatchSetExitDrag

export type MapContextType = {
    mapId: string;
    tree: GenericTree<MapTreeItem>;
    UI: {
        //
        // The Map editor can conceivably need data for:
        //    - Which tool is selected in the toolbar
        //    - Whether an exit is being dragged, from where, and to where
        // Updates to this data should be performed through the mapDispatch
        // function.
        //
        toolSelected: ToolSelected;
        exitDrag: MapContextExitDrag;
    },
    // mapD3: MapDThree,
    mapDispatch: (action: MapDispatchAction) => void;
}
