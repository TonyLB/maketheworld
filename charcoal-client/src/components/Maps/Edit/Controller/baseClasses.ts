import { SchemaConditionTag, SchemaExitTag, SchemaRoomTag } from "@tonylb/mtw-wml/dist/simpleSchema/baseClasses"

export type ToolSelected = 'Select' | 'Move' | 'AddRoom' | 'OneWayExit' | 'TwoWayExit'

export type MapTreeExit = SchemaExitTag & { inherited?: boolean }
export type MapTreeRoom = SchemaRoomTag & { inherited?: boolean; contents: MapTreeItem[] }
export type MapTreeCondition = SchemaConditionTag & { inherited?: boolean; contents: MapTreeItem[] }

export type MapTreeItem = MapTreeExit | MapTreeRoom | MapTreeCondition
