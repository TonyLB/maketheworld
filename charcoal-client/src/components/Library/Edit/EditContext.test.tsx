import renderer from 'react-test-renderer'
import React, { FunctionComponent, useEffect } from 'react'

import { useEditContext, EditSchema, EditSubListSchema } from './EditContext'
import { schemaOutputToString } from '@tonylb/mtw-wml/dist/schema/utils/schemaOutput/schemaOutputToString'
import { treeTypeGuard } from '@tonylb/mtw-wml/dist/tree/filter'
import { isSchemaOutputTag, SchemaTag } from '@tonylb/mtw-wml/dist/schema/baseClasses'
import { maybeGenericIDFromTree, stripIDFromTree } from '@tonylb/mtw-wml/dist/tree/genericIDTree'
import { GenericTree } from '@tonylb/mtw-wml/dist/tree/baseClasses'

const Render: FunctionComponent<{}> = () => {
    const { value } = useEditContext()
    return <React.Fragment>
        { schemaOutputToString(treeTypeGuard({ tree: value, typeGuard: isSchemaOutputTag })) }
    </React.Fragment>
}

describe('EditSchema', () => {

    it('should provide value for children', () => {
        expect(renderer
            .create(
                <EditSchema
                    field={{ data: { tag: 'String', value: 'Test' }, children: [], id: '' }}
                    value={[{ data: { tag: 'String', value: 'Test' }, children: [] }]}
                    onChange={() => {}}
                    onDelete={() => {}}
                >
                    <Render />
                </EditSchema>
            ).toJSON()
        ).toMatchSnapshot()

    })
})

describe('EditSubListSchema', () => {
    const testSchema: GenericTree<SchemaTag> = [
        { data: { tag: 'String', value: 'Test1' }, children: [] },
        { data: { tag: 'String', value: 'Test2' }, children: [] },
        { data: { tag: 'String', value: 'Test3' }, children: [] }
    ]

    it('should extract an indexed value from node children', () => {
        expect(renderer
            .create(
                <EditSchema
                    field={maybeGenericIDFromTree(testSchema)[0]}
                    value={testSchema}
                    onChange={() => {}}
                    onDelete={() => {}}
                >
                    <EditSubListSchema index={1}>
                        <Render />
                    </EditSubListSchema>
                </EditSchema>
            ).toJSON()
        ).toMatchSnapshot()
    })

    it('should bubble up onChange events', () => {
        const ChangeRender: FunctionComponent<{}> = () => {
            const { onChange } = useEditContext()
            useEffect(() => {
                onChange([{ data: { tag: 'String', value: 'Test change' }, children: [], id: '' }])
            }, [])
            return <Render />
        }
        const onChange = jest.fn()
        renderer.act(() => {
            renderer.create(
                <EditSchema
                    field={maybeGenericIDFromTree(testSchema)[0]}
                    value={testSchema}
                    onChange={onChange}
                    onDelete={() => {}}
                >
                    <EditSubListSchema index={1}>
                        <ChangeRender />
                    </EditSubListSchema>
                </EditSchema>
            )
        })
        expect(onChange).toHaveBeenCalledTimes(1)
        expect(stripIDFromTree(onChange.mock.calls[0][0])).toEqual([
            { data: { tag: 'String', value: 'Test1' }, children: [] },
            { data: { tag: 'String', value: 'Test change' }, children: [] },
            { data: { tag: 'String', value: 'Test3' }, children: [] }
        ])
    })

    it('should bubble up onDelete events', () => {
        const ChangeRender: FunctionComponent<{}> = () => {
            const { onDelete } = useEditContext()
            useEffect(() => {
                onDelete()
            }, [])
            return <Render />
        }
        const onChange = jest.fn()
        renderer.act(() => {
            renderer.create(
                <EditSchema
                    field={maybeGenericIDFromTree(testSchema)[0]}
                    value={testSchema}
                    onChange={onChange}
                    onDelete={() => {}}
                >
                    <EditSubListSchema index={1}>
                        <ChangeRender />
                    </EditSubListSchema>
                </EditSchema>
            )
        })
        expect(onChange).toHaveBeenCalledTimes(1)
        expect(stripIDFromTree(onChange.mock.calls[0][0])).toEqual([
            { data: { tag: 'String', value: 'Test1' }, children: [] },
            { data: { tag: 'String', value: 'Test3' }, children: [] }
        ])
    })
})