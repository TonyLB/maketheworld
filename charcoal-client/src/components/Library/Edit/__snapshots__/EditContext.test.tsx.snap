// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSchema should provide value for children 1`] = `"Test"`;

exports[`EditSubListSchema should extract an indexed value from node children 1`] = `"Test2"`;
